gadgets: Define metadata format (#2107)
inspektor-gadget/inspektor-gadget: docs/devel/hello-world-gadget.md

---
title: Hello world gadget
weight: 100
description: >
  Hello world gadget
---

> ⚠️ This feature is experimental. In order for the commands to work, the `IG_EXPERIMENTAL` env var must be set to `true`
>
> ```bash
> $ export IG_EXPERIMENTAL=true
> ```

This is a short getting started guide to write your first gadget. This guide will get you familiar
with the key concepts by implementing a simplified version of the "trace open" (opensnoop) tool.

The first step is to create an empty folder where the source code of the gadget will be stored:

```bash
$ mkdir mygadget
```

## Implementing the eBPF program

The eBPF code contains the source code for the programs that are injected in the kernel to collect
information. Let's create a file called `program.bpf.c` and put the following contents in there.

The first thing we need is to include some header files.

```c
#include <vmlinux.h>
#include <bpf/bpf_helpers.h>

#include <gadget/macros.h>
```

- `vmlinux.h`: kernel types definitions
- `bpf/bpf_helpers.h`: eBPF helpers signatures

> - `gadget/macros.h`: Inspektor Gadget macros

Author: I suppose you meant to move that as comments on the code block above, I did that.

Then, we have to specify a structure with all the information our gadget will provide. Let's only
put the pid for the time being.

```c
struct event {
	__u32 pid;
};

// we need this to make sure the compiler doesn't remove our struct
const struct event *unusedevent __attribute__((unused));
```

> Then, create a perf ring buffer eBPF map to send events to user space:

Author: opts, the map used is actually a "perf event array", not a ring buffer. I updated the sentence to make it clear. Unfortunately I can't find a similar documentation for that one.

$ mkdir mygadget
```

## Implementing the eBPF program

The eBPF code contains the source code for the programs that are injected in the kernel to collect
information. Let's create a file called `program.bpf.c` and put the following contents in there.

The first thing we need is to include some header files.

```c
// Kernel types definitions
// Check https://blog.aquasec.com/vmlinux.h-ebpf-programs for more details
#include <vmlinux.h>

// eBPF helpers signatures
// Check https://man7.org/linux/man-pages/man7/bpf-helpers.7.html to learn
// more about different available helpers
#include <bpf/bpf_helpers.h>

// Inspektor Gadget macros
#include <gadget/macros.h>
```

Then, we have to specify a structure with all the information our gadget will provide. Let's only
put the pid for the time being.

```c
struct event {
	__u32 pid;
};

// we need this to make sure the compiler doesn't remove our struct
const struct event *unusedevent __attribute__((unused));
```

Then, create a perf ring buffer eBPF map to send events to user space:

```c
struct {
	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
	__uint(key_size, sizeof(u32));
	__type(value, struct event);
} events SEC(".maps");
```

> And mark this map as a tracer map, i.e. a map that is used to push events to user space:

Author: I'll ignore this comment for now. I remember we have the same discussion a couple of years ago about tracers vs snoopers, etc. and it seems we'll having that again soon :D.

                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1113
                                                                               1219
                                                                               220121
```

> The tool only shows the PID, that's not very useful.

Author: Why not? :D

filename is being trimmed. The metadata file contains extra information about the gadget, among
other things, it can be used to specify the format to be used.

An initial version of the metadata file can be created by passing `--update-metadata` to the build command:

```bash
$ sudo -E ig image build . -t mygadget --update-metadata
```

It'll create a `gadget.yaml` file:

```yaml
name: 'TODO: Fill the gadget name'
description: 'TODO: Fill the gadget description'
tracers:
  events:
    mapName: events
    structName: event
structs:
  event:
    fields:
    - name: pid
      description: 'TODO: Fill field description'
      attributes:
        width: 16
        alignment: left
        ellipsis: end
    - name: comm
      description: 'TODO: Fill field description'
      attributes:
        width: 16
        alignment: left
        ellipsis: end
    - name: filename
      description: 'TODO: Fill field description'
      attributes:
        width: 16
        alignment: left
        ellipsis: end
```

Let's edit the file to customize the output. We define some templates for well-known fields like

> pid, comm, etc.

Author: I agree. I was even thinking that we could create specific data types for them, so the developer doesn't have to worry about setting the template, but it could be too much coupling...

                             224708          firefox-bin     /usr/lib/firefox/x86_64/x86_64/libmozsandbox.so
                             224708          firefox-bin     /usr/lib/firefox/x86_64/libmozsandbox.so
                             224708          firefox-bin     /usr/lib/firefox/x86_64/libmozsandbox.so
                             224708          firefox-bin     /usr/lib/firefox/libmozsandbox.so
```

Now the output is much better.

### Filtering and container enrichement

The gadget we created provides information about all events happenning on the host, however it (a)
doesn't provide any information about the container generating the even nor (b) allows to filter
events by a given container.

Inspektor Gadget provides the logic to filter and enrich events with container information.
This first step is to include these two addional header files

```c
#include <gadget/mntns_filter.h>
#include <gadget/types.h>
```

- `gadget/mntns_filter.h`: Defines an eBPF map and some helper functions used to filter events by containers
- `gadget/types.h`: Inspektor Gadget specific types, like `mnt_ns_id_t`.

Add the following field to the event structure.

```c

> 	mnt_ns_id_t mntns_id;

Reviewer: ```suggestion
struct event {
	...
	mnt_ns_id_t mntns_id;
	...
};
```

INFO[0000] Experimental features enabled
RUNTIME.CONTAINERNAME        PID             COMM            FILENAME
```

Nothing is shown because now the gadget is filtering only events generated by containers, create a
container and run some commands there:

```bash
$ docker run --rm -ti --name=mycontainer busybox cat /dev/null
```

Only events generated in containers are now printed, and they include the name of the container
generating them.

```bash
RUNTIME.CONTAINERNAME        PID             COMM            FILENAME
...
mycontainer                  225805          cat             /lib/tls/libm.so.6
mycontainer                  225805          cat             /lib/x86_64/x86_64/libm.so.6
mycontainer                  225805          cat             /lib/x86_64/libm.so.6
mycontainer                  225805          cat             /lib/x86_64/libm.so.6
mycontainer                  225805          cat             /lib/libm.so.6
mycontainer                  225805          cat             /lib/libresolv.so.2
mycontainer                  225805          cat             /lib/libc.so.6
mycontainer                  225805          cat             /dev/null
```

> This is also possible to filter by container name now, the following command doesn't show any event
> as there are not containers with that name.

Reviewer: ```suggestion
It's now also possible to filter by container name. The following command doesn't show any event
as there is no container with the specified name.
```

```bash
$ sudo -E ig run mygadget:latest -c non_existing_container
INFO[0000] Experimental features enabled
RUNTIME.CONTAINERNAME        PID             COMM            FILENAME
```

## Updating the gadget

Some times we have to update our gadgets, like adding more fields to the generated event for instance.
Let's add the following fields to the event struct:

```c

> 	__u32 uid;
> 	__u32 gid;

Reviewer: ```suggestion
struct event {
	...
	__u32 uid;
	__u32 gid;
	...
};
```

```

and add the logic in the eBPF program to fill them:

```c
	__u64 uid_gid = bpf_get_current_uid_gid();
	event.uid = (__u32)uid_gid;
	event.gid = (__u32)(uid_gid >> 32);
```

> Let's build the gadget with the `--update-metadata` file, so new fields are taken into
> consideration. Notice the -v option is used to get debugging messages.

Reviewer: ```suggestion
Let's build the gadget with the `--update-metadata` file, so our new fields are automatically
added to the metadata file. Notice the -v option is used to get debugging messages.
```

```c
// program.bpf.c

// Kernel types definitions
// Check https://blog.aquasec.com/vmlinux.h-ebpf-programs for more details
#include <vmlinux.h>

// eBPF helpers signatures
// Check https://man7.org/linux/man-pages/man7/bpf-helpers.7.html to learn
// more about different available helpers
#include <bpf/bpf_helpers.h>

// Inspektor Gadget macros
#include <gadget/macros.h>

struct event {
	__u32 pid;
};

// we need to "use" this struct for the compiler not to remove it in an optimization pass
const struct event *unusedevent __attribute__((unused));

struct {
	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
	__uint(key_size, sizeof(u32));
	__type(value, struct event);
} events SEC(".maps");

GADGET_TRACE_MAP(events);

SEC("tracepoint/syscalls/sys_enter_openat")
int enter_openat(struct trace_event_raw_sys_enter *ctx)
{
	struct event event = {};

	event.pid = bpf_get_current_pid_tgid() >> 32;

	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
			      &event, sizeof(event));

	return 0;
}

char LICENSE[] SEC("license") = "GPL";
```

## Building the gadget for the first time

We can now compile our gadget. You don't need to have any dependency on the machine, the `image

> build` by default uses a container with all dependencies to compile the code.

Reviewer 2: Detail to add:
> `ig` does not create the container itself but connects to the Docker socket.

Successfully built docker.io/library/mygadget:latest@sha256:dd3f5c357983bb863ef86942e36f4c851933eec4b32ba65ee375acb1c514f628
```

## (Optional) Pushing the gadget image to a container registry

You could push the gadget to a remote container registry. If you're using the same machine for
building and running the gadget, this step can be skipped.

```bash
$ sudo -E ig image tag mygadget:latest ghcr.io/mauriciovasquezbernal/mygadget:latest
INFO[0000] Experimental features enabled
Successfully tagged with ghcr.io/mauriciovasquezbernal/mygadget:latest@sha256:dd3f5c357983bb863ef86942e36f4c851933eec4b32ba65ee375acb1c514f628

$ sudo -E ig image push ghcr.io/mauriciovasquezbernal/mygadget:latest
INFO[0000] Experimental features enabled
Pushing ghcr.io/mauriciovasquezbernal/mygadget:latest...
Successfully pushed ghcr.io/mauriciovasquezbernal/mygadget:latest@sha256:dd3f5c357983bb863ef86942e36f4c851933eec4b32ba65ee375acb1c514f628
```
## Running the gadget

We're now all set to run our gadget for the first time.

```bash
$ sudo -E ig run mygadget:latest
INFO[0000] Experimental features enabled
RUNTIME.CONTAINERNAME                                                          PID

> WARN[0000] The gadget doesn't provide metadata

Reviewer 2: Should the warning message include an hint to use `--update-metadata` (actionable error message)?

Author: This message is print at run time, there is nothing a user can do as `--update-metadata` is available at build time.
